### PR TITLE
Support HTML textarea input

### PR DESCRIPTION
**Pull request by dtorres-sf**
Refactor that pulls a lot of the functionality from main.js into an input.js classes that can be used to handle more than one input type.

There is a single interface for the following functions:
  - `hasCodeElements`
  - `hasInputs`
  - `destroy`
  - `disableInput`
  - `enableInput`
  - `getValue`
  - `editorChanged`

Main.js class into `CodeInputs` instance passing in an ID. Based on the ID prefix `CodeInputs` knows whether it is a call for a CodeMirror instance or Textarea instance. 

It is easy to add more types of inputs by inheriting from `ResurfaceInput` and implementing the required functions and defining a prefix.

Quick and dirty block diagram:
```
  ┌─────────────────────────────────────────────────────────────────────┐
  │   Main.js                                                           │
  │                                                                     │
  │                                                                     │
  │                                                                     │
  └───────────────────────────────────┬──────▲──────────────────────────┘
                                      │      │
                                      │      │
                                      │      │
┌─────────────────────────────────────┼──────┼──────────────────────────────┐
│                                     │      │                              │
│  CodeInputs          ┌──────────────▼──────┴───────┐                      │
│                      │Lookup ResurfaceInput        │                      │
│                      │by ID prefix                 │                      │
│                      └──┬─────▲───────────────▲─┬──┘                      │
│                         │     │               │ │                         │
│       ┌─────────────────▼─────┴───┐   ┌───────┴─▼─────────────────┐       │
│       │ ResurfaceCodeMirrorInput  │   │ ResurfaceTextareaInput    │       │
│       │                           │   │                           │       │
│       └───────────────────────────┘   └───────────────────────────┘       │
│                                                                           │
└───────────────────────────────────────────────────────────────────────────┘
```

---

TODO:
- [x] Update textarea to get value
- [x] Update button style to work for both textarea overlay and codemirror overlay
- [x] Rename "inputs" to "targets"
- [x] Rename usage of `mirrorId` to `targetId`
- [ ] Determine if we should replace Classes with Functional style
- [ ] Add extension option to enable textareas on specific sites (using regex url checks)

---

This PR was originally opened [here](https://github.com/dtorres-sf/resurface/pull/1). 

Conversation copied over from other PR:

From lokimckay

> Thanks again for this Dustin 👍 Looks like a great start to me
>
> I'm trying to decide whether it makes sense to use Javascript classes here or if I should stick with a functional style
I don't have much experience with classes, but they do seem like they might keep `main.js` a bit cleaner
> 
> I also have a suggestion around terminology, but I'm not sure if I'm misunderstanding some parts:
> 
> Does "target" make more sense than "input" when referring to a DOM elements that can be commandeered by Resurface? E.g. should we rename `CodeInputs` -> `ResurfaceTargets`?
> 
> I've also given you collaborator access over on the [base repository](https://github.com/code-cabana/resurface/tree/support-html-inputs) 😄 

From dtorres-sf:

> lokimckay Thanks for taking a look! The reason I went with classes was to allow for the ability to quickly add more targets by simply creating a new class that inherits from the base class and overriding the necessary functions required to implement the API.
> 
> I am sure this could be easily achieved using a functional style too, and I can try that approach out. I wanted to avoid to many if/else or switch/case statements to determine what to do based on the target type. 
> 
> If we go with a functional style do you think we should mimic something similar to what I did with classes? For instance:
> 
> ```javascript
> const ResurfaceCodeMirrorInput = () => {
>   const elements = [];
> 
>   return {
>      getValue: () => {...},
>      hasInputs: () => {...},
>      ...
>   }
> }
> ```
> 
> Regarding naming I think "targets" is much better than inputs - thanks for the suggestion. I will make this update.
> 
> I plan to work on this over the weekend to try and finish some of these changes off and get textareas fully functional. I appreciate the collaborator access! I will go ahead and make changes on the branch you created.

### CODE REVIEW

**Comment by dtorres-sf** [comment]
@lokimckay I opened the PR here instead and have done some cleanup, renaming and finished the functionality for textareas. 

For now if we are going to keep the input detection the same I still need to add an allowlist of regex urls to enable for textareas for so that it is not enabled on all sites. In the future this may not be needed if a scan is triggered via the extension button.

File: browser-extension/src/content-scripts/proxy.js [hunk 1]
```diff
@@ -86,11 +86,11 @@ async function listen(message) {
       });
 
       window.__RESURFACE__.port.onDisconnect.addListener(() => {
-        debug(`Port for mirror ${mirrorId} disconnected`);
-        postMessageToDom({ type: "portDisconnected", mirrorId });
+        debug(`Port for mirror ${targetId} disconnected`);
```
Comment (lokimckay) [inline browser-extension/src/content-scripts/proxy.js:89]:
```suggestion
        debug(`Port for target ${targetId} disconnected`);
```

File: browser-extension/src/lib/sentry.js [hunk 2]
```diff
@@ -2,13 +2,13 @@ import { init, captureException } from "@sentry/browser";
...
-  init({
-    dsn: "https://5dcc5b13db974a6ba5c850d90a18c565@o1271690.ingest.sentry.io/6464314",
-    integrations: [new BrowserTracing()],
-    tracesSampleRate: 1.0,
-  });
+  //init({
+  //  dsn: "https://5dcc5b13db974a6ba5c850d90a18c565@o1271690.ingest.sentry.io/6464314",
+  //  integrations: [new BrowserTracing()],
+  //  tracesSampleRate: 1.0,
+  //});
```
Comment (lokimckay) [inline browser-extension/src/lib/sentry.js:9]:
```suggestion
  init({
    dsn: "https://5dcc5b13db974a6ba5c850d90a18c565@o1271690.ingest.sentry.io/6464314",
    integrations: [new BrowserTracing()],
    tracesSampleRate: 1.0,
  });
```

File: browser-extension/src/lib/sentry.js [hunk 3]
```diff
@@ -2,13 +2,13 @@ import { init, captureException } from "@sentry/browser";
...
 }
 
 export function sentryError(error) {
-  captureException(error);
+  //captureException(error);
```
Comment (lokimckay) [inline browser-extension/src/lib/sentry.js:13]:
```suggestion
  captureException(error);
```

File: browser-extension/src/content-scripts/main.js [hunk 4]
```diff
@@ -15,26 +13,17 @@ const error = (...args) =>
...
-  allButtons.forEach((button) => {
-    const mirrorId = button.dataset.resurfaceMirrorId;
-    postMessageToProxy({ type: "closeEditor", mirrorId, recipient: "editor" });
-    button.remove();
-  });
-  buttons = [];
-  codeMirrors = [];
+  if(resurfaceTargets){
+    resurfaceTargets.destroy();
+  }
```
Comment (lokimckay) [inline browser-extension/src/content-scripts/main.js:26]:
```suggestion
  if(resurfaceTargets) resurfaceTargets.destroy();
```

**Review by lokimckay — CHANGES_REQUESTED** [review]
Minor code style fix suggestions and re-enabling sentry

**Review by lokimckay — APPROVED** [review]

**Comment by dtorres-sf** [comment]
Thanks for the clean up - my mistake committing that commented out sentry file!

**Review by lokimckay — APPROVED** [review]

**Comment by dtorres-sf** [comment]
@lokimckay Thanks for taking a look and the cleanup/refactor! I am happy to help with the update to have user initiated action. If you want to still show the buttons when the user initiates the action - we may need to add back the absolute positioning. Testing the relative positioning on textareas leads to the button sometimes ending up in weird positions (could even be off the screen). I tried to test the absolute positioning refactor I did on a variety of different target types but I can fix it up more if you were seeing issues with that change.

**Comment by lokimckay** [comment]
Ah, I should have tested with some other sites my bad

The absolute positioning was giving me trouble on [this W3 page](https://www.w3schools.com/tags/tryit.asp?filename=tryhtml_textarea)

I don't think it's valid HTML to have a button be a child of a textarea, but if we put the button as a sibling to the textarea - it becomes a bit harder to position right?. I guess we will have to temporarily wrap the textarea in a div container that has position: relative, and insert the button as a child of that element maybe?

I'll give you access to the 1.4.x branch if you want to make any changes for this there

I'm partway through the removing polling change at the moment, I'll push the new branch tonight with some notes to show the progress :)

**Comment by dtorres-sf** [comment]
@lokimckay It is not valid HTML to make it a child in a text area. So what my changes did was make it a sibling, switch to absolute positioning, and then find the position to the closest positioned element and offset from there. It seemed to work well for both CodeMirror and textarea. I would be happy to put those changes back in the 1.4.x branch and I will make sure to test on the W3 page. Glad to hear you are making progress on removing polling!

**Comment by lokimckay** [comment]
Awesome, thats way more elegant 👍

**Pull request merged**